File: TicketChose.cpp
#include "TicketChose.hpp"
#include "Trip.hpp"
#include "Order.hpp"

using namespace InfSystBusStation;
using namespace System;
using namespace System::Collections::Generic;

// Конструктор
TicketChose::TicketChose(Trip^ trip, Order^ order)
    : selectedTrip(trip), currentOrder(order) {

    if (trip == nullptr) {
        throw gcnew ArgumentNullException("Поездка не может быть null!");
    }

    if (order == nullptr) {
        throw gcnew ArgumentNullException("Заказ не может быть null!");
    }

    availableSeats = gcnew List<int>();
    selectedSeats = gcnew List<int>();

    // Получаем количество мест из автобуса
    if (trip->GetBus() != nullptr) {
        totalSeats = trip->GetBus()->GetSeatCount(); // Нужно добавить метод GetSeatCount() в Bus
    }
    else {
        totalSeats = 40; // Значение по умолчанию
    }

    LoadAvailableSeats();
}

// Загрузить свободные места
void TicketChose::LoadAvailableSeats() {
    availableSeats->Clear();

    // 1. Все места от 1 до totalSeats
    for (int i = 1; i <= totalSeats; i++) {
        availableSeats->Add(i);
    }

    // 2. Удалить уже занятые места (нужно получить из Order или базы данных)
    // Это заглушка - нужно будет реализовать проверку занятых мест
    // Например: if (IsSeatOccupied(i)) availableSeats->Remove(i);

    // 3. Сортировка
    availableSeats->Sort();
}

// Проверить свободно ли место
bool TicketChose::IsSeatAvailable(int seatNumber) {
    return availableSeats->Contains(seatNumber);
}

// Выбрать место
bool TicketChose::SelectSeat(int seatNumber) {
    if (!IsSeatAvailable(seatNumber)) {
        return false;
    }

    if (selectedSeats->Contains(seatNumber)) {
        return false; // Уже выбрано
    }

    availableSeats->Remove(seatNumber);
    selectedSeats->Add(seatNumber);
    selectedSeats->Sort();

    return true;
}

// Отменить выбор места
bool TicketChose::DeselectSeat(int seatNumber) {
    if (!selectedSeats->Contains(seatNumber)) {
        return false;
    }

    selectedSeats->Remove(seatNumber);
    availableSeats->Add(seatNumber);
    availableSeats->Sort();

    return true;
}

// Очистить выбор
void TicketChose::ClearSelection() {
    // Возвращаем все выбранные места в доступные
    for each (int seat in selectedSeats) {
        availableSeats->Add(seat);
    }
    availableSeats->Sort();
    selectedSeats->Clear();
}

// Добавить выбранные билеты в заказ
bool TicketChose::AddSelectedTicketsToOrder() {
    if (!HasSelectedSeats()) {
        return false;
    }

    // Для каждого выбранного места создаем билет
    for each (int seat in selectedSeats) {
        // Создаем билет (нужен класс Ticket)
        // Ticket^ ticket = gcnew Ticket(selectedTrip, seat, selectedTrip->GetPrice());
        // currentOrder->AddTicket(ticket);
    }

    // Очищаем выбор после добавления
    selectedSeats->Clear();

    return true;
}

// Получить информацию о поездке
String^ TicketChose::GetTripInfo() {
    if (selectedTrip == nullptr) {
        return "Поездка не выбрана";
    }

    return String::Format(
        "МАРШРУТ: {0}\n"
        "ДАТА: {1:dd.MM.yyyy}\n"
        "ВРЕМЯ: {2}\n"
        "АВТОБУС: {3}\n"
        "ВОДИТЕЛЬ: {4}\n"
        "ЦЕНА БИЛЕТА: {5} руб.",
        selectedTrip->GetRoute(),
        selectedTrip->GetTripDate(),
        selectedTrip->GetTripTime(),
        selectedTrip->GetBus() != nullptr ?
        selectedTrip->GetBus()->GetBrand() + " " + selectedTrip->GetBus()->GetModel() : "Не указан",
        selectedTrip->GetDriver() != nullptr ?
        selectedTrip->GetDriver()->GetFullName() : "Не назначен",
        selectedTrip->GetPrice()
    );
}

// Вспомогательные методы
bool TicketChose::HasAvailableSeats() {
    return availableSeats->Count > 0;
}

bool TicketChose::HasSelectedSeats() {
    return selectedSeats->Count > 0;
}

bool TicketChose::CanAddToOrder() {
    return HasSelectedSeats() && currentOrder != nullptr;
}

int TicketChose::GetSelectedSeatsCount() {
    return selectedSeats->Count;
}

Decimal TicketChose::GetTotalPrice() {
    if (selectedTrip == nullptr) return 0;
    return selectedTrip->GetPrice() * selectedSeats->Count;
}

File: TicketChose.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TicketChose {
    private Trip currentTrip;
    private List<Integer> availableSeats;

    public TicketChose(Trip trip) {
        try {
            if (trip == null) {
                throw new IllegalArgumentException("Trip cannot be null!");
            }
            if (trip.getBus() == null) {
                throw new IllegalArgumentException("Trip must have a bus assigned!");
            }
            
            this.currentTrip = trip;
            this.availableSeats = new ArrayList<>();
            
            // Заполняем доступные места
            int totalSeats = trip.getBus().getPlaces();
            if (totalSeats <= 0) {
                throw new IllegalArgumentException("Bus must have positive number of seats!");
            }
            
            for (int i = 1; i <= totalSeats; i++) {
                this.availableSeats.add(i);
            }
            
        } catch (IllegalArgumentException e) {
            System.err.println("Error creating TicketChose: " + e.getMessage());
            throw e;
        }
    }
    

    public void showAvailableSeats() {
        try {
            if (this.currentTrip == null) {
                throw new IllegalStateException("No trip assigned!");
            }
            
            System.out.println("=== Available Seats ===");
            System.out.println("Bus: " + this.currentTrip.getBus().getBrand() 
                + " [" + this.currentTrip.getBus().getCode() + "]");      
            System.out.println("Total seats: " + this.currentTrip.getBus().getPlaces()  
                + ", available: " + this.availableSeats.size() + "\n");

            if (this.availableSeats.isEmpty()) {
                System.out.println("No available seats!");
            } else {
                for (int seat : this.availableSeats) {
                    System.out.println("Seat " + seat + " - Available");
                }
            }
            System.out.println("=======================");
            
        } catch (IllegalStateException e) {
            System.err.println("Error showing available seats: " + e.getMessage());
        }
    }

    public boolean isSeatAvailable(int seatNumber) {
        try {
            if (seatNumber <= 0) {
                throw new IllegalArgumentException("Seat number must be positive!");
            }
            
            return this.availableSeats.contains(seatNumber);
            
        } catch (IllegalArgumentException e) {
            System.err.println("Error checking seat availability: " + e.getMessage());
            return false;
        }
    }

    public boolean reserveSeat(int seatNumber) {
        try {
            if (seatNumber <= 0) {
                throw new IllegalArgumentException("Seat number must be positive!");
            }
            
            if (!this.availableSeats.contains(seatNumber)) {
                return false;  
            }
            this.availableSeats.remove(Integer.valueOf(seatNumber));
            return true;
            
        } catch (IllegalArgumentException e) {
            System.err.println("Error reserving seat: " + e.getMessage());
            return false;
        }
    }

    public void releaseSeat(int seatNumber) {
        try {
            if (seatNumber <= 0) {
                throw new IllegalArgumentException("Seat number must be positive!");
            }
            
            if (!this.isSeatAvailable(seatNumber)) {
                this.availableSeats.add(seatNumber);
                Collections.sort(this.availableSeats);
            }
            
        } catch (IllegalArgumentException e) {
            System.err.println("Error releasing seat: " + e.getMessage());
        }
    }

    public void ticketToOrder(Order order, int seatNumber, Passenger passenger, Ticket.TicketType ticketType) {
        try {
            if (order == null) {
                throw new IllegalArgumentException("Order cannot be null!");
            }
            if (passenger == null) {
                throw new IllegalArgumentException("Passenger cannot be null!");
            }
            if (seatNumber <= 0) {
                throw new IllegalArgumentException("Seat number must be positive!");
            }
            if (ticketType == null) {
                throw new IllegalArgumentException("Ticket type cannot be null!");
            }
            
            // проверка доступности места
            if (!this.availableSeats.contains(seatNumber)) {
                throw new IllegalStateException("Seat " + seatNumber + " is already taken or doesn't exist!");
            }

            // Удаление места из доступных
            this.availableSeats.remove(Integer.valueOf(seatNumber));

            // создание нового билета
            Ticket newTicket = new Ticket(seatNumber, this.currentTrip, passenger, ticketType);

            // Добавление билета в заказ
            order.addTicket(newTicket);

            System.out.println("Ticket booked!");
            System.out.println("Seat: " + seatNumber);
            System.out.println("Type: " + newTicket.getTicketTypeName());
            System.out.println("Price: " + newTicket.getFinalPrice() + " rub.");
            System.out.println("Status: " + newTicket.getStatus());
            
        } catch (IllegalArgumentException | IllegalStateException e) {
            System.err.println("Error booking ticket: " + e.getMessage());
            throw e;
        }
    }

    // Геттеры
    public List<Integer> getAvailableSeats() {
        return this.availableSeats;
    }

    public Trip getCurrentTrip() {
        return this.currentTrip;
    }
}